signer/src/stacks/contracts.rs

```diff
@@ -987,11 +1056,98 @@ impl AsContractCall for RejectWithdrawalV1 {
     ///    an event on the canonical Stacks blockchain.
     /// 2. That the signer bitmap matches the signer decisions stored in
     ///    this signer's database.
-    async fn validate<C>(&self, _ctx: &C, _req_ctx: &ReqContext) -> Result<(), Error>
+    async fn validate<C>(&self, ctx: &C, req_ctx: &ReqContext) -> Result<(), Error>
     where
         C: Context + Send + Sync,
     {
-        // TODO(255): Add validation implementation
+        if self.deployer != req_ctx.deployer {
+            return Err(WithdrawalRejectErrorMsg::DeployerMismatch.into_error(req_ctx, self));
+        }
+
+        // 1. The request exists. Check whether the associated withdrawal request transaction
+        //    is confirmed on the canonical stacks blockchain. Fail the withdrawal request if
+        //    it is not on the canonical stacks blockchain.
+        //
+        // 2. The double spend check. Check whether the qualified request ID is in the
+        //    bitcoin_withdrawals_outputs table, and that any of the associated txids are
+        //    confirmed on the canonical bitcoin blockchain. Fail the withdrawal request if
+        //    such a transaction was found.
+
+        let stacks_chain_tip = ctx
+            .get_storage()
+            .get_stacks_chain_tip(&req_ctx.chain_tip.block_hash)
+            .await
+            .unwrap()
+            .unwrap();
+        let maybe_report = ctx
+            .get_storage()
+            .get_withdrawal_request_report(
+                &req_ctx.chain_tip.block_hash,
+                &stacks_chain_tip.block_hash,
+                &self.id,
+                &ctx.config().signer.public_key(),
+            )
+            .await?;
+
+        let Some(report) = maybe_report else {
+            return Err(WithdrawalRejectErrorMsg::RequestMissing.into_error(req_ctx, self));
+        };
+
+        match report.status {
+            WithdrawalRequestStatus::Confirmed => (),
+            WithdrawalRequestStatus::Fulfilled(_txid) => {
+                // fails #2
+                return Err(WithdrawalRejectErrorMsg::RequestFulfilled.into_error(req_ctx, self));
+            }
+            WithdrawalRequestStatus::Unconfirmed => {
+                // fails #1
+                return Err(WithdrawalRejectErrorMsg::RequestUnconfirmed.into_error(req_ctx, self));
+            }
+        }
+
+        // 3. The request is final. Check that we've seen six new bitcoin blocks since observing
+        //    the bitcoin anchor block associated with the Stacks block confirming the withdrawal
+        //    request transaction. Fail the withdrawal request if we’ve observed less than 6
+        //    bitcoin blocks since the anchor block.
+
+        let request_block_height = report.bitcoin_block_height;
+        let blocks_observed = req_ctx.chain_tip.block_height - request_block_height;
+
+        if blocks_observed < 6 {
+            return Err(WithdrawalRejectErrorMsg::RequestNotFinal.into_error(req_ctx, self));
+        }
```

Comment:
In https://github.com/stacks-network/sbtc/issues/1363 we only care about `WITHDRAWAL_BLOCKS_EXPIRY` being observed so we can remove this as well.

Reply:
Done